cmds/modprobe/modprobe.go

```diff
@@ -0,0 +1,170 @@
+
+	scanner := bufio.NewScanner(rd)
+	for scanner.Scan() {
+		txt := scanner.Text()
+		nameDeps := strings.Split(txt, ":")
+		modPath, modDeps := nameDeps[0], nameDeps[1]
+		modPath = strings.TrimSpace(modPath)
+
+		var dependency dependency
+		if len(modDeps) > 0 {
+			for _, dep := range strings.Split(strings.TrimSpace(modDeps), " ") {
+				dependency.deps = append(dependency.deps, dep)
+			}
+		}
+		deps[modPath] = &dependency
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return deps, nil
+}
+
+func findModPath(modName string, m depMap) (string, error) {
+	for mp := range m {
+		if strings.HasSuffix(mp, modName+".ko") {
+			return mp, nil
+		}
+	}
+
+	return "", fmt.Errorf("modprobe: Could not find path for module %s", modName)
+}
+
+func loadModule(modPath, options string) error {
+	filename := filepath.Join(moduleDir, modPath)
+	f, err := os.Open(filename)
+	if err != nil {
+		return fmt.Errorf("modprobe: Could not open %q: %v", filename, err)
+	}
+	defer f.Close()
+
+	if err, e := kmodule.FileInit(f, options, 0); err != nil && e != unix.EEXIST {
+		return fmt.Errorf("modprobe: could not load %q: %v", filename, err)
+	}
+
+	return nil
+}
+
+func loadDeps(modPath string, m depMap) error {
+	dependency, ok := m[modPath]
+	if !ok {
+		return fmt.Errorf("modprobe: Could not find dependency %s", modPath)
+	}
+
+	if dependency.state == loading {
+		return fmt.Errorf("modprobe: Circular dependency, %s already in state LOADING", modPath)
+	} else if dependency.state == loaded {
+		return nil
+	}
+
+	m[modPath].state = loading
+
+	for _, dep := range dependency.deps {
+		if err := loadDeps(dep, m); err != nil {
+			return err
+		}
+	}
+
+	// done with dependencies, load module
+	if err := loadModule(modPath, ""); err != nil {
+		return err
+	}
+	m[modPath].state = loaded
+
+	return nil
+}
+
+func main() {
+	if len(os.Args) == 1 {
+		log.Fatalf("Usage: %s", cmd)
+	}
+
+	rel, err := exec.Command("uname", "-r").Output()
+	if err != nil {
+		log.Fatal(err)
+	}
+	moduleDir = filepath.Join(moduleDir, strings.TrimSpace(string(rel)))
+
+	f, err := os.Open(filepath.Join(moduleDir, "modules.dep"))
+	if err != nil {
+		log.Fatalf("modprobe: Could not open dependency file: %v", err)
+	}
+	defer f.Close()
+
+	deps, err := genDeps(f)
+	if err != nil {
+		log.Fatalf("modprobe: Could not generate dependency map: %v", err)
+	}
+
+	modName := os.Args[1]
+
+	modPath, err := findModPath(modName, deps)
+	if err != nil {
+		log.Fatalf("modprobe: Could not find module path %s: %v", modName, err)
+	}
+
+	// The strategy is to load all dependencies (without options), and then load the final toplevel module with
+	// options. IMO this is a bit nicer than to start with the toplevel module (which we could) and do "level
+	// tracking" to find out when we want to use module options.
+	deps[modPath].state = loading // could be a very weird dep cycle
+	for _, d := range deps[modPath].deps {
+		if err := loadDeps(d, deps); err != nil {
```

Comment:
This is a completely fabricated case anyways. If you fabricated modules.dep (or depmod did it), then your system is horribly broken anyways. That check is there to not run in circles, I don't think this will ever happen on a real system. You would have to generate maliciously crafted modules + dependencies. If your system is that broken, loading some  then unused modules is your least concern. So I don't really see a need to change that.

Reply:
Reflecting a bit more on that I really see no improvement. You want (no you really need a module), in one case it tells you that your system is broken beyond repair, in the other case it loads some modules and tells you your system is beyond repair. So...

Reply:
I don't think it's acceptable to leave the system in a broken state, even if the user fucked up. "This won't ever happen" are famous last words...

(I recently had to fabricate module stuff for complicated reasons and of course, the first couple times I fucked it up. It can happen..)

Reply:
Yes, that is why the code handles it. It loads all the dependencies and the module or not, like the regular `modprobe`. In very weird cases (but they are handled), it loaded some dependencies before loading then next dependency fails. But I don't see a "broken system state" at all, at least not one modprobe caused. And checking for dep-loops first does improve that particular case, but loading later dependencies can fail for various reasons. What should we do if we decided that everyting is happy and then the last dependency fails to load (for what ever reasons). Then we should clean up the already loaded modules? To me that really does not make sense. That is additional code that IMO does not help at all.

Again, in very weird situations (hat are handled) we might end up not loading all dependencies and tell the user. Why would some loaded modules hurt? In the end you *started* with a broken system, modprobe did not make it worse (if you like me don't consider a handfull unused loaded modules as the introduction of a "broken system state").

Reply:
Alright, point taken. You're right.

Reply:
For what it is worth, Linux's modprobe creates a list of dependencies before inserting the first one. See this function: https://git.kernel.org/pub/scm/utils/kernel/kmod/kmod.git/tree/libkmod/libkmod-module.c#n1274

That being said, this is probably not worth the time to fix. It only helps when `modules.dep` is incorrectly constructed.